File: src/harness/pricing.py
from __future__ import annotations

from dataclasses import dataclass


@dataclass(frozen=True)
class ModelPricing:
    input_per_mtok: float
    output_per_mtok: float


PRICING: dict[str, ModelPricing] = {
    "claude-opus-4-7": ModelPricing(15.00, 75.00),
    "claude-opus-4-7-1m": ModelPricing(30.00, 150.00),
    "opus": ModelPricing(15.00, 75.00),
    "opus[1m]": ModelPricing(30.00, 150.00),
    "claude-sonnet-4-6": ModelPricing(3.00, 15.00),
    "sonnet": ModelPricing(3.00, 15.00),
    "sonnet[1m]": ModelPricing(6.00, 22.50),
    "claude-haiku-4-5": ModelPricing(1.00, 5.00),
    "haiku": ModelPricing(1.00, 5.00),
    "gpt-5.4": ModelPricing(3.00, 12.00),
    "gpt-5.4-mini": ModelPricing(0.40, 1.60),
    "gpt-5.4-high": ModelPricing(3.00, 12.00),
    "gpt-5.3-codex": ModelPricing(3.00, 12.00),
    "gpt-5.3-codex-spark": ModelPricing(0.40, 1.60),
    "gpt-5.5": ModelPricing(5.00, 20.00),
    "gpt-4.1": ModelPricing(2.00, 8.00),
    "o3": ModelPricing(2.00, 8.00),
    "gemini-2.5-pro": ModelPricing(1.25, 10.00),
    "gemini-2.5-flash": ModelPricing(0.30, 2.50),
    "qwen3-coder-plus": ModelPricing(1.00, 4.00),
    "qwen3-coder-flash": ModelPricing(0.20, 0.80),
}

FAMILY_PREFIXES: list[tuple[str, ModelPricing]] = [
    ("gpt-5.4", ModelPricing(3.00, 12.00)),
    ("gpt-5.5", ModelPricing(5.00, 20.00)),
    ("gpt-4", ModelPricing(2.00, 8.00)),
    ("gpt-5", ModelPricing(3.00, 12.00)),
    ("gemini", ModelPricing(1.25, 10.00)),
    ("qwen", ModelPricing(1.00, 4.00)),
    ("opus", ModelPricing(15.00, 75.00)),
    ("sonnet", ModelPricing(3.00, 15.00)),
    ("haiku", ModelPricing(1.00, 5.00)),
    ("claude", ModelPricing(3.00, 15.00)),
]


def lookup_pricing(model: str | None) -> ModelPricing | None:
    if not model:
        return None
    idx = model.find("/")
    ident = model[idx + 1 :] if idx >= 0 else model
    lower = ident.lower().strip()

    if ident in PRICING:
        return PRICING[ident]
    if lower in PRICING:
        return PRICING[lower]

    for prefix, pricing in FAMILY_PREFIXES:
        if lower.startswith(prefix):
            return pricing
    return None


def derive_cost(model: str | None, tokens_in: int | None, tokens_out: int | None) -> float | None:
    pricing = lookup_pricing(model)
    if pricing is None:
        return None
    if tokens_in is None and tokens_out is None:
        return None
    in_cost = (tokens_in or 0) / 1_000_000 * pricing.input_per_mtok
    out_cost = (tokens_out or 0) / 1_000_000 * pricing.output_per_mtok
    return round(in_cost + out_cost, 6)

File: ts/src/pricing.ts
// Per-token pricing for models that don't emit cost in their session output.
// USD per 1M tokens. Prices last verified 2026-04-25 against vendor docs.
// Adapters whose CLIs don't report costUsd (codex, gemini, qwen, aider) use
// these to compute cost from tokensIn/tokensOut, so SessionTelemetry.costUsd
// is non-null whenever tokens are non-null.

export interface ModelPricing {
  /** USD per 1M input tokens. */
  inputPerMTok: number
  /** USD per 1M output tokens. */
  outputPerMTok: number
}

// Index keyed by canonical model id. Match priority: exact id, then prefix, then family.
const PRICING: Record<string, ModelPricing> = {
  // Anthropic
  'claude-opus-4-7':            { inputPerMTok: 15.00, outputPerMTok: 75.00 },
  'claude-opus-4-7-1m':         { inputPerMTok: 30.00, outputPerMTok: 150.00 },
  'opus':                       { inputPerMTok: 15.00, outputPerMTok: 75.00 },
  'opus[1m]':                   { inputPerMTok: 30.00, outputPerMTok: 150.00 },
  'claude-sonnet-4-6':          { inputPerMTok:  3.00, outputPerMTok: 15.00 },
  'sonnet':                     { inputPerMTok:  3.00, outputPerMTok: 15.00 },
  'sonnet[1m]':                 { inputPerMTok:  6.00, outputPerMTok: 22.50 },
  'claude-haiku-4-5':           { inputPerMTok:  1.00, outputPerMTok:  5.00 },
  'haiku':                      { inputPerMTok:  1.00, outputPerMTok:  5.00 },

  // OpenAI / Codex
  'gpt-5.4':                    { inputPerMTok:  3.00, outputPerMTok: 12.00 },
  'gpt-5.4-mini':               { inputPerMTok:  0.40, outputPerMTok:  1.60 },
  'gpt-5.4-high':               { inputPerMTok:  3.00, outputPerMTok: 12.00 },
  'gpt-5.3-codex':              { inputPerMTok:  3.00, outputPerMTok: 12.00 },
  'gpt-5.3-codex-spark':        { inputPerMTok:  0.40, outputPerMTok:  1.60 },
  'gpt-5.5':                    { inputPerMTok:  5.00, outputPerMTok: 20.00 },
  'gpt-4.1':                    { inputPerMTok:  2.00, outputPerMTok:  8.00 },
  'o3':                         { inputPerMTok:  2.00, outputPerMTok:  8.00 },

  // Google
  'gemini-2.5-pro':             { inputPerMTok:  1.25, outputPerMTok: 10.00 },
  'gemini-2.5-flash':           { inputPerMTok:  0.30, outputPerMTok:  2.50 },

  // Qwen (Alibaba)
  'qwen3-coder-plus':           { inputPerMTok:  1.00, outputPerMTok:  4.00 },
  'qwen3-coder-flash':          { inputPerMTok:  0.20, outputPerMTok:  0.80 },
}

const FAMILY_PREFIXES: { prefix: string; pricing: ModelPricing }[] = [
  { prefix: 'gpt-5.4',  pricing: { inputPerMTok: 3.00, outputPerMTok: 12.00 } },
  { prefix: 'gpt-5.5',  pricing: { inputPerMTok: 5.00, outputPerMTok: 20.00 } },
  { prefix: 'gpt-4',    pricing: { inputPerMTok: 2.00, outputPerMTok:  8.00 } },
  { prefix: 'gpt-5',    pricing: { inputPerMTok: 3.00, outputPerMTok: 12.00 } },
  { prefix: 'gemini',   pricing: { inputPerMTok: 1.25, outputPerMTok: 10.00 } },
  { prefix: 'qwen',     pricing: { inputPerMTok: 1.00, outputPerMTok:  4.00 } },
  { prefix: 'opus',     pricing: { inputPerMTok: 15.00, outputPerMTok: 75.00 } },
  { prefix: 'sonnet',   pricing: { inputPerMTok:  3.00, outputPerMTok: 15.00 } },
  { prefix: 'haiku',    pricing: { inputPerMTok:  1.00, outputPerMTok:  5.00 } },
  { prefix: 'claude',   pricing: { inputPerMTok:  3.00, outputPerMTok: 15.00 } },
]

export function lookupPricing(model: string | null | undefined): ModelPricing | null {
  if (!model) return null
  // Strip provider prefix if present (anthropic/sonnet → sonnet)
  const idx = model.indexOf('/')
  const id = idx >= 0 ? model.slice(idx + 1) : model
  const lower = id.toLowerCase().trim()

  if (PRICING[id]) return PRICING[id]
  if (PRICING[lower]) return PRICING[lower]

  for (const { prefix, pricing } of FAMILY_PREFIXES) {
    if (lower.startsWith(prefix)) return pricing
  }
  return null
}

export function deriveCost(model: string | null, tokensIn: number | null, tokensOut: number | null): number | null {
  const pricing = lookupPricing(model)
  if (!pricing) return null
  if (tokensIn === null && tokensOut === null) return null
  const inCost = (tokensIn ?? 0) / 1_000_000 * pricing.inputPerMTok
  const outCost = (tokensOut ?? 0) / 1_000_000 * pricing.outputPerMTok
  return Number((inCost + outCost).toFixed(6))
}
